libraries/UR_Atmosphere/UR_Atmosphere.cpp

```diff
@@ -0,0 +1,222 @@
+   Copyright (c) 2019 Hiroshi Takey <htakey@gmail.com>
+
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+/*
+   UPDATE 05072019:
+
+   The improvements about QNH, QFE, Namespace ISA_MATH_CONST Definitions constants and
+   Altitude_difference calculations was based on the AviationCalculator 1.8.2 05/10/2017
+   revision application from Ph.D. Aerospace Engineering, Joachim K. Hochwarth (GE Principal
+   Engineer/System Architect), published on this URL:
+
+   http://www.hochwarth.com/misc/AviationCalculator.html
+   http://www.hochwarth.com/script/AviationCalculator.js
+
+   Thanks Hochwarth!
+
+   These functions was tested on the Encoder & Altimeter Benchtest Hardware and Application
+   development in cooperation and support with Fenix Aeronautics Academy under the supervision
+   and technical support with the FAA Licensed Fixed and Rotor Wing Aircraft Engineer
+   Maintenance Service, Rudy Averanga <raveranga@fenixaeronautica.com>.
+
+   Thanks a lot Rudy!
+*/
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Baro/AP_Baro.h>
+#include "UR_Atmosphere.h"
+
+extern const AP_HAL::HAL& hal;
+bool UR_Atmosphere::_initialized = false;
+UR_Atmosphere *UR_Atmosphere::_instance = nullptr;
+
+UR_Atmosphere::UR_Atmosphere()
+{
+    _barometer = AP_Baro::get_singleton();
+}
+
+void UR_Atmosphere::init(void)
+{
+    if (_initialized) {
+        return;
+    }
+
+    _initialized = true;
+    _instance = this;
+
+    if (!_barometer) {
+        _barometer = new AP_Baro;
+        _barometer->init();
+    }
+}
+
+void UR_Atmosphere::update(uint8_t sensor)
+{
+    if (!atm_status.updated) {
+        atm_status.updated = true;
+        atm_status.altitude_amsl = calculate_altitude_difference(press0, _barometer->sensors[sensor].pressure, _barometer->sensors[sensor].temperature);
+        atm_status.qnh_pressure = calculate_qnh(atm_status.altitude_amsl, (_barometer->sensors[sensor].pressure / 100), _barometer->sensors[sensor].temperature, ALTITUDE_UNIT::METER);
+        atm_status.qfe_pressure = calculate_qfe(atm_status.altitude_amsl, atm_status.qnh_pressure, _barometer->sensors[sensor].temperature, ALTITUDE_UNIT::METER) * 100;
+    }
+}
+
+// Get the calculation altitude difference based on th Layer 0 > Troposphere.
+float UR_Atmosphere::calculate_altitude_difference(float base_pressure, float pressure, float temp, altitude_unit_t alt_diff_unit) const
```

Comment:
This produces very nearly the same result we already get, I don't see what this does for you, except break badly if you provide a unit other then meters.

Reply:
Check please the above comment to understand it.